### DISCUSSION TITLE
How to prevent bazel from downloading JDK11 and JDK21?

### DISCUSSION THREAD

**Discussion by jason-sachs** [bazelbuild/bazel#21752] [Q&A] [answered]
We're using JDK8u152 for licensing reasons, but when I follow the Java tutorial (https://bazel.build/start/java) and set JAVA_HOME to point to my installed JDK, when I `bazel build //:ProjectRunner` it downloads a copy of JDK11 and JDK21... or at least there are around 640MB in `c:\users\{MY_USER_NAME}\_bazel_{MY_USER_NAME}\vwhg3rug\external\` under `rules_java~~toolchains~remotejdk11_win` and `rules_java~~toolchains~remotejdk21_win`.

How do I prevent this?

**Comment by hvadehra** [comment, marked as answer]
Yes.

Forcing the use of **only** the local jdk  involves two things:

1. The default **target** `java_runtime` is `local_jdk` but the default **execution** `java_runtime` is `remotejdk_11`. To use the local jdk for execution, use `--tool_java_runtime_version=local_jdk`

2. If a project does nothing wrt java toolchains in `MODULE.bazel` / `WORKSPACE`, the `rules_java` order of registration will be followed where we register the remote jdks _before_ the local one, so those "win" by default. The way to avoid this is to explicitly register the local jdk toolchain first for the project (in `MODULE.bazel` / `WORKSPACE`). Alternatively, use `--extra_toolchains=@local_jdk//:all`

**Reply by stagnation** [reply]
[For the record, sorry to hijack your thread, but it does not seem like a top-level solution either.] Another technique to avoid downloading the built-in repos is to register a stub-repo earlier. This blocks the download of the JDKs with a custom repository rule that takes the name and does nothing*, so if it is ever downloaded there will be no data transfer, and you get errors trying to use things. This works for projects that do not use Java at all (and not the coverage report code).

\* May need some target names that would be loaded, which can probably be done with a do-nothing genrule or something.

**Reply by t-rad679** [reply]
When I use `--extra_toolchains=@local_jdk//:all` Bazel tells me that it can't find the repository `local_jdk`. Any idea what's going on?

**Reply by hvadehra** [reply]
@t-rad679 Please consider filing a bug at https://github.com/bazelbuild/rules_java/issues/new/choose with more details (a minimal reproducer would be ideal).
